### DISCUSSION TITLE
How to handle battery packs?

### DISCUSSION THREAD

**Discussion by cyasing** [pybamm-team/PyBaMM#2612] [Q&A] [answered]
I am trying to simulate a battery pack for an electric vehicle. The battery pack needs to be subjected to a drive cycle of one of WLTC drive cycles. I am not able to understand how to use an array of cell models to a single drive cycle (instead of subjecting each cell to a drive cycle). This is to ensure that I am not just repeating the same simulation over and over, but is simultaneous as in real-life conditions.

I do see that there is a discussion for WLTC drive cycle here, but it is scaled for a single cell, which is not something that I want.

**Comment by cyasing** [comment]
It looks like there is another software developed by the PyBaMM team, liionpack, that uses PyBaMM to model a whole battery pack. I'll be trying that now.

**Reply by cyasing** [reply]
It says in liionpack that there is no support for variable current drive cycles, so I'm going to have to do something else now.

**Reply by TomTranter** [reply]
You can use a drive cycle with liionpack. https://liionpack.readthedocs.io/en/latest/examples/05%20Drive%20cycles/

**Reply by cyasing** [reply]
@TomTranter  I edited my single cell script and tried it, but it gave this error:

![image](https://user-images.githubusercontent.com/59028917/215063893-87f0f0cf-540f-4d9c-bb9d-1ea2cf5fd462.png)

Hence, I thought that was not supported.

Also, I tried the script in the link, it solves it, but doesn't show any plots. I tried the drive cycle example given in the liionpack source, its plots appear and vanish in an instant. I have no idea why.

**Reply by TomTranter** [reply]
Can you post your script that's failing?

**Reply by cyasing** [reply]
Yes, certainly. I've attached them here. It should have both the script as well as the data.
[test.zip](https://github.com/pybamm-team/PyBaMM/files/10517980/test.zip)

**Reply by cyasing** [reply]
@TomTranter I redid liionpack installation, and the plots are appearing just fine now. My script however, still gives the same error.

**Reply by cyasing** [reply]
@TomTranter Did you find anything wrong with my approach? I have now updated my single cell script, data, and similarly, the battery script that works better than the previous one, at least at the cell level. The error of "constant current operations" still does not go away, however. Just saying if those are needed.

**Reply by TomTranter** [reply]
Sorry haven't looked yet. Will see on Monday

**Reply by cyasing** [reply]
> Sorry haven't looked yet. Will see on Monday

Alright.

**Reply by TomTranter** [reply]
So liionpack supports drivecycles specifying current only

**Reply by cyasing** [reply]
> So liionpack supports drivecycles specifying current only

Yeah, that makes sense.

**Comment by TomTranter** [comment, marked as answer]
Aligning liionpack more closely with PyBaMM functionality is on our roadmap and should be available in the near future. Sorry for the inconvenience. In the meantime that error message isn't the best as I see how it could have led you to think no drive cycles were possible at all

**Reply by cyasing** [reply]
I see. It's fine, I hope the team works it out well. In the meantime, can you recommend any way in which I can get the power drive cycle simulated for a battery pack using PyBaMM, with or without liionpack? A useful reference? If it's not too much, of course.

**Reply by TomTranter** [reply]
The solving process in liionpack is very different from PyBaMM and we're basically working to completely change the way the problem is set up and defined so the PyBaMM solvers can be used directly. In the mean time you could just divide power by the voltage at previous step or by some nominal voltage which wouldn't be too hard in liionpack. I think getting this to work in liionpack wouldn't be too tricky if you can figure it out yourself I can give pointers but can't spend any time on it myself right now

**Reply by cyasing** [reply]
Alright, I'll check this out. Thanks!

**Reply by cyasing** [reply]
Can you tell how to access the values of P, V, and I correctly for corresponding time? I think I'm missing something.

**Reply by TomTranter** [reply]
Have a look at the model.variables. You're going to have to be a bit more specific with that question. At which point are you trying to access them?

**Reply by cyasing** [reply]
I was planning on editing the protocols.py to add a case for power input. In a forked repository. I thought that might be better than writing a function in my original script, since I can stop that error from coming up that way.

**Reply by cyasing** [reply]
> I was planning on editing the protocols.py to add a case for power input. In a forked repository. I thought that might be better than writing a function in my original script, since I can stop that error from coming up that way.

Do you think this is not optimal?

**Reply by TomTranter** [reply]
Please stop posting here and open an issue on the liionpack repo.

**Reply by cyasing** [reply]
Sorry about that, I thought I should continue here.
